src/main/java/net/md_5/specialsource/Jar.java

```diff
@@ -75,12 +75,26 @@
      * @return
      */
     public boolean containsClass(String clazz) {
-        try {
-            return contains.contains(clazz) ? true : getClass(clazz) != null;
-        } catch (IOException ex) {
-            // IO error - regardless we do not have access to the class, so we can ignore it and move on
-            return false;
+        if (contains.contains(clazz)) {
+            return true;
+        }
+
+        if (containsResource(clazz + ".class")) {
+            contains.add(clazz);
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Check if this jar contains the given resource.
+     *
+     * @param name The name of the resource
+     * @return true Whether a resource with the given name exists
+     */
+    public boolean containsResource(String name) {
+        JarFile jarFile = jarForResource.get(name);
+        return jarFile != null && jarFile.getEntry(name) != null;
```

Comment:
The `&& jarFile.getEntry(name) != null` check is probably redundant, it mirrors the check in `getResource`.

Not sure whether or not you want this.